feat: add support for s11 from library

YogevBokobza opened the pull request:
<!-- markdownlint-disable MD041-->
## Description

BREAKING CHANGE: all routes now require a _type_ query param.

## Checklist

- [x] I have followed this repository's contributing guidelines.
- [x] I will adhere to the project's code of conduct.

## Additional information

> Anything else?

YogevBokobza committed:
feat: add support for s11 from library

YogevBokobza committed:
fix: versionsing

YogevBokobza commented:
@dmatik
Can you test that everything is working?

For my tests it does.

TomerFi commented:
@dmatik I will do my best to review this PR over the weekend, but I can't test it. Your help with this would be much appreciated. 😄

dmatik commented:
> @dmatik I will do my best to review this PR over the weekend, but I can't test it. Your help with this would be much appreciated. 😄 

Yes,
I saw the PR, was just on vacation with the Family. 
Will surely test it once promoted.

YogevBokobza committed:
fix: tests and linting

YogevBokobza committed:
fix: test_web_app.py

TomerFi commented:
![image](https://github.com/user-attachments/assets/4a982a2d-3fed-4aad-afed-a69f61976a88)

@YogevBokobza 
I assume the [python:3.11.4-slim](https://github.com/TomerFi/switcher_webapi/blob/dev/Dockerfile#L1) base image for [linux/arm/v7](https://github.com/TomerFi/switcher_webapi/blob/dev/.github/workflows/pr.yml#L93) is missing C libs required for building _pycryptodome_.

We had this problem before with other dependencies. Luckily, these supported a workaround, note the *x_NO_EXTENSIONS* environments variables we use with our [install](https://github.com/TomerFi/switcher_webapi/blob/dev/Dockerfile#L13) command. See [aiohttp](https://github.com/search?q=repo%3Aaio-libs%2Faiohttp%20NO_EXTENSIONS&type=code), [frozenlist](https://github.com/search?q=repo%3Aaio-libs%2Ffrozenlist%20_NO_EXTENSIONS&type=code), [multidict](https://github.com/search?q=repo%3Aaio-libs%2Fmultidict%20NO_EXTENSIONS&type=code), and [yarl](https://github.com/search?q=repo%3Aaio-libs%2Fyarl%20NO_EXTENSIONS&type=code). Unfortunately, from a quick look in [pycryptodome](https://github.com/Legrandin/pycryptodome), I wonder if a pure-python build is supported.

TomerFi commented:
Based on the description from both [pycryptodome](https://pypi.org/project/pycryptodome/) and [pycryptodomex](https://pypi.org/project/pycryptodomex/), I assume a pure Python build doesn't exist:

```text
To the largest possible extent, algorithms are implemented in pure Python.
Only the pieces that are extremely critical to performance (e.g. block ciphers) are implemented as C extensions.
```

YogevBokobza commented:
> Based on the description from both [pycryptodome](https://pypi.org/project/pycryptodome/) and [pycryptodomex](https://pypi.org/project/pycryptodomex/), I assume a pure Python build doesn't exist:
> 
> ```
> To the largest possible extent, algorithms are implemented in pure Python.
> Only the pieces that are extremely critical to performance (e.g. block ciphers) are implemented as C extensions.
> ```

So.. You have any idea how to solve this?

TomerFi commented:
We can probably solve this by building outside the image, using a different base image or a build state, or installing the required extensions. But this probably means this PR is blocked until we figure out how to install pycryptodome in arm for RPIs.
I'll mark this PR as a draft to avoid accidentally merging it. I'll open an issue (#717) referencing this PR.

Commit:
build: install build essentials for building pycryptodome

Signed-off-by: Tomer Figenblat <tomer@tomfi.info>

TomerFi commented:
@YogevBokobza I'm trying to push a fix, but I can't get the CI trigger; I think I can workaround this by temporarily patching the CI.
Please don't push or anything, I'll keep you updated.

TomerFi commented:
@YogevBokobza Alright, everything works now.
I'm marking this as ready for review; please update your local branch before working on it further.

TomerFi commented on requirements_test.txt:
```diff
@@ -1,3 +1,4 @@
+requests==2.32.3
```
Do we need the same dependency for both testing and runtime?

YogevBokobza replied:
No.. I remove that.

TomerFi commented on requirements.txt:
```diff
@@ -1,2 +1,3 @@
-aioswitcher==3.4.2
+aioswitcher==4.0.0
```
Let's change this to the new 4.0.2.

YogevBokobza replied:
Updated.

TomerFi commented on requirements.txt:
```diff
@@ -1,2 +1,3 @@
-aioswitcher==3.4.2
+aioswitcher==4.0.0
+requests==2.32.3
```
Can we make do with _aiohttp_ instead of adding _requests_?

YogevBokobza replied:
Yes. removed requests.

YogevBokobza committed:
fix: remove requests and update aioswitcher version to 4.0.2

YogevBokobza commented:
@TomerFi 
Do you think need to add something more to inform users to go and fetch the token from the Switcher API?

YogevBokobza committed:
Revert "feat: added git to dockerfile to help debuging git inside dev container"

This reverts commit a5e34516f04eb54bd4fc5e25a1f82400eb345ac5.

YogevBokobza committed:
feat: added git to dockerfile to help debuging git inside dev container

TomerFi commented:
> @TomerFi Do you think need to add something more to inform users to go and fetch the token from the Switcher API?

Probably yes.

For the record, I haven't had the time to review this PR behind the extent of adding the missing C libs and the requests dependencies that caught my eye while browsing. I'll review this in full this week.

YogevBokobza commented:
> > @TomerFi Do you think need to add something more to inform users to go and fetch the token from the Switcher API?
> 
> Probably yes.
> 
> For the record, I haven't had the time to review this PR behind the extent of adding the missing C libs and the requests dependencies that caught my eye while browsing. I'll review this in full this week.

Sure.. If you have something in the mind about where and how to tell that let me know..
By the way, I added git to the installs in the Dockerfile as it helps debug the git inside the dev container.

TomerFi commented:
> By the way, I added git to the installs in the Dockerfile as it helps debug the git inside the dev container.

I'm not sure I understand how is git used in this context.

YogevBokobza commented:
> > By the way, I added git to the installs in the Dockerfile as it helps debug the git inside the dev container.
> 
> I'm not sure I understand how is git used in this context.

When I am using Visual Code it will reopen the project in a docker container, where I can type in commands while inside the container.

When I typed "git status" for example it said the "git command is not found"
This is hurting development inside the dev container to be able to push commits to PRs.

TomerFi commented:
How does git relate to debugging? How does 'git status' help  debugging this project?

TomerFi commented:
> When I am using Visual Code it will reopen the project in a docker container, where I can type in commands while inside the container.
> 
> When I typed "git status" for example it said the "git command is not found" This is hurting development inside the dev container to be able to push commits to PRs.

In that case, create a Dockerfile targeted for development; don't load development-only tools on the runtime-targeted Dockerfile. Regardless, this project's Dockerfile is due for refactoring for a while. I want to try to separate the build stage from the runtime stage. So this won't last.

TomerFi commented:
For the record, I think someone else wanted to add a devcontainer Dockerfile here once, but I don't remember what the end of that was; maybe we even had one at some point. I'm currently not a VSCode user, and even when I were, I disliked working inside a container. It's slow! 🐌 . I only did it when there was a good enough reason for that (I can't remember any).
Most of this project (all code) was written with VSCode and I've never seen a good enough reason to develop it in a container.

That said, if both you and Dima prefer a devcontainer, I guess we can add a designated Dockerfile.

YogevBokobza committed:
Update README.md

YogevBokobza commented:
Removed the git command and added description for the token part

TomerFi commented on README.md:
```diff
@@ -25,6 +25,11 @@
+<h4 align="center">
+For a newer devices there is a need to get a token from Switcher API. For more info see <a href="https://github.com/TomerFi/aioswitcher/wiki">here</a>
+</h4>
+
```
```suggestion
<p align="center">
Newer device types might require a token. See info <a href="https://github.com/TomerFi/aioswitcher/wiki">here</a>.
</p>
```
Also, move this further down, just above the _contributors_ section in line 71+.

YogevBokobza replied:
Done.

TomerFi commented on app/webapp.py:
```diff
@@ -313,22 +361,31 @@ async def control_breeze_device(request: web.Request) -> web.Response:
-        target_temp = int(body[KEY_TARGET_TEMP]) if body.get(KEY_TARGET_TEMP) else None
-        fan_level = thermostat_fan_levels.get(body.get(KEY_FAN_LEVEL))
-        thermostat_swing = thermostat_swings.get(body.get(KEY_THERMOSTAT_SWING))
+        device_state = device_states.get(body.get(KEY_DEVICE_STATE, None))
+        thermostat_mode = thermostat_modes.get(body.get(KEY_THERMOSTAT_MODE, None))
+        target_temp = int(body[KEY_TARGET_TEMP]) if body.get(KEY_TARGET_TEMP) else 0
+        fan_level = thermostat_fan_levels.get(body.get(KEY_FAN_LEVEL, None))
+        thermostat_swing = thermostat_swings.get(body.get(KEY_THERMOSTAT_SWING, None))
```
What's the reason for the _None_ default addition? Are we handling _None_ for these properties? Because if not, it's best to not set a default and let the code throw an error when needed.

YogevBokobza replied:
This is to match [this](https://github.com/TomerFi/aioswitcher/blob/dev/src/aioswitcher/api/__init__.py#L638-L642)
I can revert it back if you want..

TomerFi replied:
As long as it's handled. I don't mind.

TomerFi commented on app/webapp.py:
```diff
@@ -25,11 +25,15 @@
     ThermostatMode,
     ThermostatSwing,
 )
+from aioswitcher.device.tools import convert_str_to_devicetype
```
Based on the usage of this utility function and the modification to _openapi.yml_ spec, it looks like we're expecting users to pass the device type as strings in the query params:

```
Switcher Mini,
Switcher Power Plug,
Switcher Touch,
Switcher V2 (esp),
Switcher V2 (qualcomm),
Switcher V4,
Switcher Breeze,
Switcher Runner,
Switcher Runner Mini,
Switcher Runner S11
```

Is that the right approach? We should support more url-friendly values and handle the translation ourselves.

YogevBokobza replied:
Why not? this is the same approach as [here](https://github.com/TomerFi/aioswitcher/blob/dev/scripts/control_device.py)

TomerFi replied:
Python API vs URL Query Params.

For _Switcher V2 (qualcomm)_, the URL encoding will be _type=Switcher+V2+%28qualcomm%29_.
While something _type=v2qual_ will make the same sense and look much better.

Creating a target enum in this project would be much better. We can easily convert the values before sending them to aioswitcher. Something like:

```
mini,
plug,
touch,
v2esp,
v2qual,
v4,
breeze,
runner,
runnermini,
runners11
```

YogevBokobza replied:
Done.

TomerFi replied:
Okay, I looked at the new implementation. Thank you—that is a step in the right direction, but we still need to do something else.

Firstly, we don't need _aioswticher_'s _convert_str_to_devicetype_ utility function. You can remove this import.

Secondly, delete the added function _convert_url_devicetype_to_devicetype_; it's the only one that uses the abovementioned import. Instead of this function, add the following dictionary somewhere at the start of the file:
```python
DEVICES = {
    "mini": DeviceType.MINI,
    "plug": DeviceType.POWER_PLUG,
    "touch": DeviceType.TOUCH,
    "v2esp": DeviceType.V2_ESP,
    "v2qual": DeviceType.V2_QCA,
    "v4": DeviceType.V4,
    "breeze": DeviceType.BREEZE,
    "runner": DeviceType.RUNNER,
    "runnermini": DeviceType.RUNNER_MINI,
    "runners11": DeviceType.RUNNER_S11,
}
```

Next, modify all the invocations of the removed function from:
```python
device_type = convert_url_devicetype_to_devicetype(request.query[KEY_TYPE])
```
To:
```python
device_type = DEVICES[request.query[KEY_TYPE]]
```

---

This also means that if users pass an unknown key, i.e.:
```python
device_type = DEVICES["wrongdevicekey"]
```
 An error will be thrown, which is a good thing! Just make sure you're aware of that when modifying the code. In the current implementation, _mini_ is the default value.

YogevBokobza replied:
Done.

YogevBokobza committed:
fix based on requested changes

YogevBokobza committed:
Update README.md

Signed-off-by: YogevBokobza <yogevbokobza12@gmail.com>

YogevBokobza committed:
fix based on requested changes

YogevBokobza committed:
Update webapp.py

TomerFi commented:
@YogevBokobza You can take a look [here](https://github.com/TomerFi/switcher_webapi/blob/dev/CONTRIBUTING.md#code-development) for instructions on running the linters locally and fixing the linting errors. 

We also need to update the PR body with the summary of the breaking changes. i.e.:
```
BREAKING CHANGE: all routes now require a _type_ query param.
```
---
@dmatik Do you think you might find some time soon to take a look or test this before we merge and release it?

YogevBokobza committed:
fix linting

YogevBokobza commented:
> @YogevBokobza You can take a look [here](https://github.com/TomerFi/switcher_webapi/blob/dev/CONTRIBUTING.md#code-development) for instructions on running the linters locally and fixing the linting errors.
> 
> We also need to update the PR body with the summary of the breaking changes. i.e.:
> 
> ```
> BREAKING CHANGE: all routes now require a _type_ query param.
> ```
> 
> @dmatik Do you think you might find some time soon to take a look or test this before we merge and release it?

Fixed lint and updated the description.

TomerFi commented:
@YogevBokobza LGTM - Thank you very much!
I want to wait a bit more for Dima before merging this.

YogevBokobza commented:
> @YogevBokobza LGTM - Thank you very much! I want to wait a bit more for Dima before merging this.

It has been 2 months now and no response.. It also delays me in updating this repo for the Switcher Runner S12..
What do we do?

dmatik commented:
> > @YogevBokobza LGTM - Thank you very much! I want to wait a bit more for Dima before merging this.
> 
> 
> 
> It has been 2 months now and no response.. It also delays me in updating this repo for the Switcher Runner S12..
> 
> What do we do?

Hi,
Sorry I missed that part where you were waiting for me. 
You want me to test something?
I can do, but I only have Switcher Touch (V3, for Bioler). 
Let me know which version to take from docker. I will test it gladly.

YogevBokobza commented:
> > > @YogevBokobza LGTM - Thank you very much! I want to wait a bit more for Dima before merging this.
> > 
> > 
> > It has been 2 months now and no response.. It also delays me in updating this repo for the Switcher Runner S12..
> > What do we do?
> 
> Hi, Sorry I missed that part where you were waiting for me. You want me to test something? I can do, but I only have Switcher Touch (V3, for Bioler). Let me know which version to take from docker. I will test it gladly.

I guess you can test that everything is still working with your existing devices.
Let us know..

dmatik commented:
> I guess you can test that everything is still working with your existing devices. Let us know..

Sure!
Can you just advise, how to I test it?
Any specific docker image we have for testing?
Or I am supposed to build an image myself to test it in container?
I see only 1.3.0 there, which is same as "latest", 9 months ago. What am I missing?

TomerFi commented:
@dmatik 
It would be best to pull the fork and build it yourself.
If not, we can do it quicker; we can merge this, and it should be available as an _early-access_ tag you can use.

dmatik commented:
@YogevBokobza @TomerFi ,
Hi,
So I cloned Yogev's fork, and ran locally in vscode, using devcontainer.
Tested "get_sate", "turn_on" (with and without Timer) and "turn_off" on my Switcher Touch. All works as expected. 

Sorry for taking it so long, totally missed that you were waiting for me.

TomerFi commented:
Thank you very much @dmatik !! 🚀 
So, we're ok to merge?

dmatik commented:
> Thank you very much @dmatik !! 🚀 
> 
> So, we're ok to merge?

Yep

TomerFi commented:
@YogevBokobza @dmatik - Thank you both very much!

Release [2.0.0](https://github.com/TomerFi/switcher_webapi/releases/tag/2.0.0).

Things to note:
- We should add a temporary note to the README.md file indicating that the migration process from version 1.x.x to version 2.x.x requires the _type_ query param.
- We should document the available _types_ and their correlated devices from [here](https://github.com/TomerFi/switcher_webapi/blob/dev/app/webapp.py#L67), it didn't work well for the generated documentation.
